File: LeetCode/cpp/946_Validate_Stack_Sequence.cpp
class Solution {
public:
	bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
		int popIdx = 0;
		stack<int> st;

		for (int i = 0; i < pushed.size(); i++) {
			while (!st.empty() && st.top() == popped[popIdx]) {
				st.pop();
				popIdx++;
			}

			st.push(pushed[i]);
		}

		while (!st.empty() && st.top() == popped[popIdx]) {
			st.pop();
			popIdx++;
		}

		if (st.empty()) {
			return true;
		}

		return false;
	}
};

File: LeetCode/java/946_Validate_Stack_Sequence.java
import java.util.Stack;

class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        int popIdx = 0;
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < pushed.length; i++) {
            while (!stack.isEmpty() && stack.peek() == popped[popIdx]) {
                stack.pop();
                popIdx++;
            }

            stack.push(pushed[i]);
        }

        while (!stack.isEmpty() && stack.peek() == popped[popIdx]) {
            stack.pop();
            popIdx++;
        }

        if (stack.isEmpty()) {
            return true;
        }

        return false;
    }
}